src/audio/ipcgtw.c

```diff
@@ -0,0 +1,377 @@
+		    0x95, 0x87, 0x2f, 0x35, 0xff, 0x8d, 0x12, 0xe8);
+
+DECLARE_TR_CTX(ipcgtw_comp_tr, SOF_UUID(ipcgtw_comp_uuid), LOG_LEVEL_INFO);
+
+/* Host communicates with IPC gateways via global IPC messages. To address a particular
+ * IPC gateway, its node_id is sent in message payload. Hence we need to keep a list of existing
+ * IPC gateways and their node_ids to search for a gateway host wants to address.
+ */
+struct ipcgtw_data {
+	union ipc4_connector_node_id node_id;
+	struct comp_dev *dev;
+	struct list_item item;
+
+	/* IPC gateway buffer size comes in blob at creation time, we keep size here
+	 * to resize buffer later at ipcgtw_params().
+	 */
+	uint32_t buf_size;
+};
+
+/* List of existing IPC gateways */
+static struct list_item ipcgtw_list_head = { 0 };
+
+static struct comp_dev *ipcgtw_new(const struct comp_driver *drv,
+				   const struct comp_ipc_config *config,
+				   const void *spec)
+{
+	struct comp_dev *dev;
+	struct ipcgtw_data *ipcgtw_data;
+	const struct ipc4_copier_gateway_cfg *gtw_cfg = spec;
+	const struct ipc4_ipc_gateway_config_blob *blob;
+
+	comp_cl_dbg(&comp_ipcgtw, "ipcgtw_new()");
+
+	dev = comp_alloc(drv, sizeof(*dev));
+	if (!dev)
+		return NULL;
+	dev->ipc_config = *config;
+
+	ipcgtw_data = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*ipcgtw_data));
+	if (!ipcgtw_data) {
+		rfree(dev);
+		return NULL;
+	}
+
+	comp_set_drvdata(dev, ipcgtw_data);
+
+	ipcgtw_data->node_id = gtw_cfg->node_id;
+	ipcgtw_data->dev = dev;
+
+	if (!gtw_cfg->config_length) {
+		comp_cl_err(&comp_ipcgtw, "ipcgtw_new(): empty ipc4_gateway_config_data");
+		goto error;
+	}
+
+	blob = (const struct ipc4_ipc_gateway_config_blob *)
+		((const struct ipc4_gateway_config_data *)gtw_cfg->config_data)->config_blob;
+
+	/* Endpoint buffer is created in copier with size specified in copier config. That buffer
+	 * will be resized to size specified in IPC gateway blob later in ipcgtw_params().
+	 */
+	comp_cl_dbg(&comp_ipcgtw, "ipcgtw_new(): buffer_size: %u", blob->buffer_size);
+	ipcgtw_data->buf_size = blob->buffer_size;
+
+	if (!ipcgtw_list_head.next)
+		list_init(&ipcgtw_list_head);
+
+	list_init(&ipcgtw_data->item);
+	list_item_append(&ipcgtw_data->item, &ipcgtw_list_head);
+
+	dev->state = COMP_STATE_READY;
+	return dev;
+
+error:
+	rfree(ipcgtw_data);
+	rfree(dev);
+	return NULL;
+}
+
+static void ipcgtw_free(struct comp_dev *dev)
+{
+	struct ipcgtw_data *ipcgtw_data = comp_get_drvdata(dev);
+
+	list_item_del(&ipcgtw_data->item);
+
+	rfree(ipcgtw_data);
+	rfree(dev);
+}
+
+static struct comp_dev *find_ipcgtw_by_node_id(union ipc4_connector_node_id node_id)
+{
+	struct list_item *item;
+
+	list_for_item(item, &ipcgtw_list_head) {
+		struct ipcgtw_data *data = list_item(item, struct ipcgtw_data, item);
+
+		if (data->node_id.dw == node_id.dw)
+			return data->dev;
+	}
+
+	return NULL;
+}
+
+static inline void audio_stream_copy_bytes_from_linear(const void *linear_source,
```

Comment:
I think we already have such func in audio_stream.h? or Im missing something?

https://github.com/thesofproject/sof/blob/df5eaf07cdd83c0928c58dc87f00c58c3e1032c6/src/include/sof/audio/audio_stream.h#LL734C10-L734C10

Reply:
Those in audio_stream.h does work with samples, these are working with bytes. So a separate implementation was required. These new functions could be potentially moved to audio_stream.h to be reused if also required somewhere else, however, I was not sure about it and left them here.